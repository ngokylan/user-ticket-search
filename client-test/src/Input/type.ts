import {
  ChangeEventHandler,
  KeyboardEventHandler,
  FocusEventHandler,
  MouseEventHandler,
  RefObject,
} from 'react';

export type InputType = 'text' | 'email' | 'number' | 'password' | 'search';
export type InputStatus = 'error' | 'success' | 'warning';

export type InputProps = {
  /** Input change callback, triggered when value is changed */
  onChange: ChangeEventHandler<HTMLInputElement>;
  /** Input click callback */
  onClick?: MouseEventHandler<HTMLInputElement>;
  /** Input mouse down callback */
  onMouseDown?: MouseEventHandler<HTMLInputElement>;
  /** Input key down callback, triggered when user presses a key, (optional) */
  onKeyDown?: KeyboardEventHandler<HTMLInputElement>;
  /** Input enter key press callback, triggered when the 'enter' button key is pressed, (optional) */
  onEnterPress?: KeyboardEventHandler<HTMLInputElement>;
  /** Input key press call back, triggered when any button key is pressed, (optional) */
  onKeyPress?: KeyboardEventHandler<HTMLInputElement>;
  /** Input losing focus callback, triggered when the input is losing focus, (optional) */
  onBlur?: FocusEventHandler<HTMLInputElement>;
  /** Input focus callback, triggered when the input is focussed, (optional) */
  onFocus?: FocusEventHandler<HTMLInputElement>;
  /** Input statusk, use one from the following: error | success | warning, (optional) */
  status?: InputStatus;
  /** Input enabled/disabled status, (optional) */
  isDisabled?: boolean;
  /** Input readonly status, (optional) */
  isReadOnly?: boolean;
  /** Input hidden status, (optional) */
  isHidden?: boolean;
  /** Input placeholder, (optional) */
  placeholder?: string;
  /** Input htmlType, one from the following: text | email | number | password | search, (optional) */
  htmlType?: InputType;
  /** Input id attribute, (optional) */
  id?: string;
  /** Input name attribute, (optional) */
  name?: string;
  /** Input icon attribute, (optional) */
  icon?: string;
  /** Input value, one from the following type: string | number, (optional) */
  value?: string | number;
  /** Whether or not it should be auto focus  */
  autoFocus?: boolean;
  /** Input DOM reference */
  ref?: RefObject<HTMLInputElement>;
  /** tabIndex is for WCAG accessibility compliance (optional).
   * By default, `tabIndex` defaults to zero, which means the element
   * should be focusable in sequential keyboard navigation, in the
   * order defined by the document's source order.
   * It is recommended that you avoid using tabIndex values greater than 0.
   * Doing so makes it difficult for people who rely on assistive technology
   * to navigate and operate page content. Instead, write the document with
   * the elements in a logical sequence. See here for more info:
   * https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex
   */
  tabIndex?: number;
  /** ariaLabel is for WCAG accessibility compliance (optional)
   * For full WCAG compliance, you must include an `ariaLabel` whenever
   * the input field does not have an associated 'label' element.
   * Please, read links below for the explanation:
   * https://dequeuniversity.com/rules/axe/3.3/label?application=AxeChrome
   * https://getbootstrap.com/docs/4.3/components/input-group/#accessibility
   */
  ariaLabel?: string;
  /** ariaDescribedby is for WCAG accessibility compliance (optional)
   * Use for linking the input component other HTML elements (e.g span, div,
   * button, etc) that provide information about the input field,
   * by referencing the element's ID property.
   * Please, see the example and the explanation by the link below
   * https://getbootstrap.com/docs/4.3/components/input-group/#button-addons
   */
  ariaDescribedby?: string;
  /** ariaInvalid is for WCAG accessibility compliance (optional)
   * The aria-invalid attribute is used to indicate that the value entered into an
   * input field does not conform to the format expected by the application and
   * is therefore invalid. For full WCAG compliance, you should always ensure
   * this prop has a value if the input field fails during form validation.
   * If it fails, then `ariaInvalid` should be set to `true`, and the prop
   * `ariaErrorMessage` should be given a value as well. See this link for more
   * info:
   * https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute
   */
  ariaInvalid?: boolean;
  /** ariaErrorMessageId is for WCAG accessibility compliance (optional)
   * This is the ID of the element describing the error (for example div, span, etc.)
   */
  ariaErrorMessageId?: string;
};

export type InputStates = {};
