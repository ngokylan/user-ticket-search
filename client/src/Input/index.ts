export * from './Input';
export { default } from './Input';
export * from './type';
