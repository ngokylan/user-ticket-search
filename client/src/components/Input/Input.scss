@import '../../_lib/style/index';
.input {
  padding: 12px 16px;
  display: inline-block;
  border: 1px solid #ccc;
  box-shadow: 0px;
  border-radius: 4px;
  box-sizing: border-box;
  text-align: left;
  height: 48px;
  width: 100%;
  @include text-style-body-rg;
  color: $gray-700;
  transition: all 0.25s;
  &:disabled {
    background-color: $gray-300;
    color: $gray-rgba-600;
    -webkit-text-fill-color: unset;
    -webkit-opacity: 1;
  }
  &:read-only {
    background-color: $gray-300;
    -webkit-text-fill-color: unset;
    -webkit-opacity: 1;
  }
  &:focus {
    outline: 0;
    border-color: #129fea;
  }
  &:focus:invalid {
    color: #b94a48;
    border-color: #e9322d;
  }
  &::-webkit-input-placeholder {
    color: $gray-600;
  }
}

.input {
  &--on-error {
    border: 1px solid #ff9fa7;
    &:focus {
      outline: 0;
      border: 1px solid #d62735;
    }
  }
  &--on-warning {
    border: 1px solid #ffe58f;
    &:focus {
      outline: 0;
      border: 1px solid #ffd666;
    }
  }
  &--on-success {
    border: 1px solid #81e8b3;
    &:focus {
      outline: 0;
    }
  }
}
