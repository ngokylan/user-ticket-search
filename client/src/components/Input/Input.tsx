import React, { Component, createRef } from 'react';
import { getClass } from '../../_lib/helper';
import { InputProps, InputStates } from './type';
import './Input.scss';

class Input extends Component<InputProps, InputStates> {
  static defaultProps: Partial<InputProps> = {
    isReadOnly: false,
    isDisabled: false,
    isHidden: false,
    htmlType: 'text',
    tabIndex: 0,
    placeholder: ''
  };

  private inputRef = createRef<HTMLInputElement>();

  componentDidUpdate() {
    if (this.props.autoFocus && this.inputRef.current) {
      this.inputRef.current.focus();
    }
  }

  handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { isDisabled, onChange, onEnterPress } = this.props;
    if (e === undefined) {
      return;
    }

    if (isDisabled) {
      e.stopPropagation();
      return;
    }

    if ((e as any).keyCode === 13 && onEnterPress) {
      onEnterPress(e as any);
    }

    if (onChange) {
      onChange(e);
    }
  };

  handleClick = (e: React.MouseEvent<HTMLInputElement>) => {
    // TODO: Add isDisabled check
    const { onClick } = this.props;
    if (onClick) {
      onClick(e);
    }
  };

  handleOnKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { isDisabled, onEnterPress, onKeyPress } = this.props;
    if (e === undefined) {
      return;
    }

    if (isDisabled) {
      e.stopPropagation();
      return;
    }

    if (e.charCode === 13 && onEnterPress) {
      onEnterPress(e);
    }

    if (onKeyPress) {
      onKeyPress(e);
    }
  };

  handleMouseDown = (e: React.MouseEvent<HTMLInputElement>) => {
    const { onMouseDown } = this.props;
    if (onMouseDown) {
      onMouseDown(e);
    }
  };

  handleOnBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    const { onBlur } = this.props;
    if (onBlur) {
      onBlur(e);
    }
  };

  handleOnFocus = (e: React.FocusEvent<HTMLInputElement>) => {
    const { onFocus } = this.props;
    if (onFocus) {
      onFocus(e);
    }
  };

  render() {
    const {
      children,
      isDisabled,
      isReadOnly,
      isHidden,
      value,
      status,
      id,
      name,
      icon,
      htmlType,
      autoFocus,
      ref,
      placeholder,
      tabIndex,
      ariaLabel,
      ariaDescribedby,
      ariaInvalid,
      ariaErrorMessageId,
    } = this.props;

    const className: string = getClass('input', {
      [`input--on-${status}`]: status,
    });

    const ariaInvalidStatus =
      ariaInvalid || status === 'error' || status === 'warning';

    return (
      <input
        id={id}        
        data-testid={`input-${id || 'default'}`}
        className={className}
        onClick={this.handleClick}
        onChange={this.handleOnChange}
        onKeyPress={this.handleOnKeyPress}
        onBlur={this.handleOnBlur}
        onFocus={this.handleOnFocus}
        disabled={isDisabled}
        readOnly={isReadOnly}
        hidden={isHidden}
        value={value}
        type={htmlType}
        name={name}
        placeholder={placeholder}
        onMouseDown={this.handleMouseDown}
        ref={this.inputRef}
        tabIndex={tabIndex}
        aria-label={ariaLabel}
        aria-describedby={ariaDescribedby}
        aria-invalid={ariaInvalidStatus}
        aria-errormessage={
          ariaInvalidStatus && ariaErrorMessageId
            ? ariaErrorMessageId
            : undefined
        }
      />
    );
  }
}

export default Input;
