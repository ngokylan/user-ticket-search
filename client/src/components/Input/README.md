Default input
```js
<Input value="user name" onChange={() => {}} />
```

Input with warning status
```js
<Input value="user name" onChange={() => {}} status="warning" />
```

Input with error status
```js
<Input value="user name" onChange={() => {}} status="error" />
```

Input with success status
```js
<Input value="user name" onChange={() => {}} status="success" />
```

Disabled input
```js
<Input value="user name" onChange={() => {}} isDisabled />
```