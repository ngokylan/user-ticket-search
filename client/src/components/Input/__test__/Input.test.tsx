import React, { Component } from 'react';
import { mount } from 'enzyme';
import Input, { InputProps } from '..';

const defaultProps: InputProps = {
  isReadOnly: false,
  isDisabled: false,
  isHidden: false,
  htmlType: 'text',
  tabIndex: 0,
};
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Input {...props} />);
  const component = wrapper.find('.input');

  return { props, wrapper, component };
};

describe('Input', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Input.defaultProps)
        );
      });

      it.each([
        ['input-default', {}],
        ['input-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { component } = mountComponent(props);
        expect(component.prop('data-testid')).toBe(expectedTestId);
      });
    });

    describe('custom', () => {
      it.each([
        ['id', 'custom-id', { id: 'custom-id' }],
        ['name', 'custom name', { name: 'custom name' }],
        [
          'placeholder',
          'input placeholder',
          { placeholder: 'input placeholder' },
        ],
        ['value', 'some value', { value: 'some value' }],
        ['type', 'text', { type: 'text' }],
        ['isReadOnly', true, { isReadOnly: true }],
        ['isReadOnly', false, { isReadOnly: false }],
        ['hidden', true, { hidden: true }],
        ['hidden', false, { hidden: false }],
        ['ariaLabel', 'some label', { ariaLabel: 'some label' }],
        ['ariaDescribedby', 'some id', { ariaDescribedby: 'some id' }],
        ['ariaInvalid', true, { ariaInvalid: true }],
        ['ariaInvalid', false, { ariaInvalid: false }],
      ])('should render prop %s as %s for Input', (name, expected, props) => {
        const { wrapper } = mountComponent(props);
        const CheckboxInput = wrapper.find('Input');
        expect(CheckboxInput.prop(name)).toBe(expected);
      });
    });
  });

  describe('interaction', () => {
    it.each([
      ['should', '.input', 1, {}],
      ['should', '.custom-class', 1, { className: 'custom-class' }],
    ])('%s have %s', (name, element, expected, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper.find(element)).toHaveLength(expected);
    });

    it.each([
      ['error', 'input--on-error', { status: 'error' }],
      ['success', 'input--on-success', { status: 'success' }],
      ['warning', 'input--on-warning', { status: 'warning' }],
    ])('on status %s class should be %s', (status, expected, props) => {
      const { component } = mountComponent(props);
      expect(component.hasClass(expected)).toEqual(true);
    });

    it.each([
      ['disabled', 'disabled', true, { isDisabled: true }],
      ['disabled', 'enabled', false, { isDisabled: false }],
      ['hidden', 'hidden', true, { isHidden: true }],
      ['hidden', 'visible', false, { isHidden: false }],
      ['readOnly', 'readOnly', true, { isReadOnly: true }],
    ])('attr %s should be %s', (attr, desc, expected, props) => {
      const { wrapper } = mountComponent(props);
      expect(
        wrapper.findWhere(item => item.prop(attr) === expected)
      ).toHaveLength(1);
    });

    it('Should focus on componentDidUpdate', () => {
      const { wrapper } = mountComponent({ autoFocus: true });
      const { inputRef }: any = wrapper.instance();

      jest.spyOn(inputRef.current, 'focus');

      wrapper.instance().componentDidUpdate();
      expect(inputRef.current.focus).toHaveBeenCalledTimes(1);
    });

    it('Should not focus on componentDidUpdate if disabled', () => {
      const { wrapper } = mountComponent({
        isDisabled: true,
      });
      const { inputRef }: any = wrapper.instance();

      jest.spyOn(inputRef.current, 'focus');

      wrapper.instance().componentDidUpdate();
      expect(inputRef.current.focus).toHaveBeenCalledTimes(0);
    });

    it('Should return undefined in there is no event on key press ', () => {
      const { wrapper } = mountComponent();
      const res = wrapper.instance().handleOnKeyPress();
      expect(res).toBeUndefined();
    });

    it('Should not handle onKeyPress if desabled', () => {
      const mockFunction = jest.fn();

      const { wrapper } = mountComponent({
        isDisabled: true,
        onKeyPress: mockFunction,
      });

      wrapper.find('input').simulate('keypress', { key: 'l' });
      expect(mockFunction).toBeCalledTimes(0);
    });

    it('Should call keypress callback', () => {
      const mockFunction = jest.fn();

      const { wrapper } = mountComponent({
        onKeyPress: mockFunction,
      });

      wrapper.find('input').simulate('keyPress', { charCode: 13 });
      expect(mockFunction).toBeCalledTimes(1);
    });

    it('Should handle onEnterPress', () => {
      const mockFunction = jest.fn();

      const { wrapper } = mountComponent({
        onEnterPress: mockFunction,
      });

      wrapper.find('input').simulate('keyPress', { charCode: 13 });
      expect(mockFunction).toBeCalledTimes(1);
    });

    it('should have onChange', () => {
      const mockCallBack = jest.fn();
      const { wrapper } = mountComponent({
        onChange: mockCallBack,
      });
      const input = wrapper.find('input');

      input.simulate('change', { target: { value: 'Changed' } });
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('should have onChange with Enter press', () => {
      const mockCallBack = jest.fn();
      const { wrapper } = mountComponent({
        onChange: mockCallBack,
        onEnterPress: mockCallBack,
      });
      const input = wrapper.find('input');

      input.simulate('change', { target: { value: 'Changed' }, keyCode: 13 });

      expect(mockCallBack.mock.calls.length).toEqual(2);
    });

    it('should not have onChange if isDesabled', () => {
      const mockCallBack = jest.fn();
      const { wrapper } = mountComponent({
        onChange: mockCallBack,
        isDisabled: true,
      });
      const input = wrapper.find('input');

      input.simulate('change', { target: { value: 'Changed' } });
      expect(mockCallBack.mock.calls.length).toEqual(0);
    });

    it('should not have onChange if event is undefined', () => {
      const { wrapper } = mountComponent();
      const res = wrapper.instance().handleOnChange();
      expect(res).toBeUndefined();
    });

    it('should call onClick when we click on input', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent({
        onClick: mockCallBack,
      });
      component.simulate('click');
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('should call onClick even if Input is disabled', () => {
      const mockCallBack = jest.fn();
      const { wrapper } = mountComponent({
        onClick: mockCallBack,
        isDisabled: true,
      });
      const event = {};
      const input = wrapper.instance();
      input.handleClick(event);
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('should not call onClick when onClick is not provided', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent();
      component.simulate('click');
      expect(mockCallBack.mock.calls.length).toEqual(0);
    });

    it('should have mouseDown', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent({
        onMouseDown: mockCallBack,
      });
      component.simulate('mousedown');
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('should not call mousedown event when mouseDown is not provided', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent();
      component.simulate('mousedown');
      expect(mockCallBack.mock.calls.length).toEqual(0);
    });

    it('should have blur', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent({
        onBlur: mockCallBack,
      });
      component.simulate('blur');
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('should not call blur event when blur is not provided', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent();
      component.simulate('blur');
      expect(mockCallBack.mock.calls.length).toEqual(0);
    });

    it('should have focus', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent({
        onFocus: mockCallBack,
      });
      component.simulate('focus');
      expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    it('should not call focus event when focus is not provided', () => {
      const mockCallBack = jest.fn();
      const { component } = mountComponent();
      component.simulate('focus');
      expect(mockCallBack.mock.calls.length).toEqual(0);
    });
  });
});
