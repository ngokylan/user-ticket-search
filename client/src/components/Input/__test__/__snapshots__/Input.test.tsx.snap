// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input rendering should be match with snapshot 1`] = `
<input
  aria-invalid={false}
  className="input"
  data-testid="input-default"
  disabled={false}
  hidden={false}
  onBlur={[Function]}
  onChange={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  onMouseDown={[Function]}
  readOnly={false}
  tabIndex={0}
  type="text"
/>
`;
